stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/utils/extensions/Channel.kt

```diff
@@ -24,54 +27,60 @@ internal fun Channel.getDisplayName(): String = name.takeIf { it.isNotEmpty() }
     ?: getUsers().joinToString { it.name }
 
 internal fun Channel.getLastMessage(): Message? =
-    messages
-        .asSequence()
+    messages.asSequence()
         .filter { it.createdAt != null || it.createdLocallyAt != null }
         .filter { it.deletedAt == null }
         .filter { !it.silent }
         .filter { it.type == ModelType.message_regular }
         .maxByOrNull { it.createdAt ?: it.createdLocallyAt!! }
 
-internal fun Channel.getCurrentUserLastMessage(): Message? = getLastMessageByUserId(getCurrentUser().id)
-
-internal fun Channel.getLastMessageByUserId(userId: String): Message? =
-    messages.lastOrNull {
-        it.user.id == userId &&
-            !it.isFailed() &&
-            !it.isDeleted() &&
-            !it.isInThread() &&
-            !it.isEphemeral()
-    }
-
-internal fun Channel.getUnreadUsers(): List<ChannelUserRead> =
-    read.filter { it.lastRead?.before(lastMessageAt) == true }
-
-internal fun Channel.getLastMessageReadCount(): Int =
-    read.filter { userRead ->
-        lastMessageAt?.let { lastMessage ->
-            userRead.lastRead?.before(lastMessage)
-        } == false
-    }.count()
-
-internal fun Channel.getLastMessageTime(): Date? = getLastMessage()?.let { it.createdAt ?: it.createdLocallyAt }
-
 internal fun Channel.getCurrentUser(): User = ChatDomain.instance().currentUser
 
-internal fun Channel.getCurrentUserRead(): ChannelUserRead? =
-    read.firstOrNull { it.user.id == getCurrentUser().id }
-
 internal fun Channel.diff(other: Channel): ChannelDiff =
     ChannelDiff(
         nameChanged = name != other.name,
         avatarViewChanged = getUsers() != other.getUsers(),
-        readStateChanged = unreadCount != other.unreadCount,
+        readStateChanged = read != other.read,
         lastMessageChanged = getLastMessage() != other.getLastMessage()
     )
 
-internal fun Channel.lastMessageByCurrentUserWasRead(): Boolean {
-    return getCurrentUserLastMessage()?.let { currentUserLastMessage ->
-        read.any { channelRead ->
-            channelRead.lastRead?.before(currentUserLastMessage.createdAt)?.not() ?: false
-        }
-    } ?: false
+internal fun Channel.isMessageRead(message: Message): Boolean =
+    read.filterNot { it.user.isCurrentUser() }
+        .mapNotNull { it.lastRead }
+        .any { it.time >= message.getCreationDate().time }
+
+internal fun Channel.isDirectMessaging() =
+    memberCount == 2 && members.map { it.user.id }.any { it == getCurrentUser().id }
```

Comment:
If we're doing perf improvements and changing this line, I'd also fetch `getCurrentUser()` just once, outside the loop

Reply:
something like `internal val currentUser by lazy { ChatDomain.instance().currentUser }`? Would this cause issues if the user switches accounts? 🤔

Reply:
I guess it wouldn't since it's per channel instance.

Reply:
Wow, I never considered delegates being used in extension properties before. This should be illegal. Like literally, because it's super broken and nobody will have the right expectations for it, I'm shocked that the compiler allows it 😄 

It simply ignores that it's an extension, it will have a single delegate backing it. But it looks extremely confusing.

---

I'd just do this for this function:

```kotlin
internal fun Channel.isDirectMessaging(): Boolean {
    val currentUser = getCurrentUser()
    return memberCount == 2 && members.map { it.user.id }.any { it == currentUser.id }
}
```

Reply:
[Example of how bad this is](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS40LjAiLCJwbGF0Zm9ybSI6ImphdmEiLCJhcmdzIjoiIiwianNDb2RlIjoiIiwibm9uZU1hcmtlcnMiOnRydWUsInRoZW1lIjoiaWRlYSIsImNvZGUiOiJcbmltcG9ydCBrb3RsaW4ucmFuZG9tLlJhbmRvbVxuXG5jbGFzcyBVc2VyXG5cbnZhbCBVc2VyLmlkOiBMb25nIGJ5IGxhenkgeyBSYW5kb20ubmV4dExvbmcoKSB9XG5cbmZ1biBtYWluKCkge1xuICAgIHZhbCB1c2VyMSA9IFVzZXIoKVxuICAgIHByaW50bG4odXNlcjEuaWQpXG4gICAgXG4gICAgdmFsIHVzZXIyID0gVXNlcigpXG4gICAgcHJpbnRsbih1c2VyMi5pZClcbn0ifQ==)

Reply:
Interesting - just saw this